File: clonegraph.js
//Objective is to make a deep copy of a graph.


let map = new Map()

function dfs(node) {
    if (!node) {
        return null
    }

    //If the node was already visited, return the clone from the visited map
    let newNode = new Node(node.val)
    if (map.has(node)) {
        newNode = node
    }

    //Set a key value pair of the original and the clone
    map.set(node, newNode)

    //Copy all the neighbors into the map, if the neighbor doesn't exist
    //yet in the map dfs through the neighbor to add all of its neighbors
    //and itself into the map
    for (let neighbor of node.neighbors) {
        newNode.neighbors.push(map.get(neighbor) || dfs(neighbor))
    }

    return newNode
}
return dfs(node)

File: clonegraph.java
//Java Solution

class Solution {
    public Node cloneGraph(Node node) {
        Map<Node, Node> graph = new HashMap<>();
        return dfs(graph, node);
    }
    
    public Node dfs(Map<Node, Node> graph, Node node) {
        if (node == null) {
            return null;
        }
        
        Node newNode = new Node(node.val);
        if (graph.containsKey(node)) {
            newNode = node;
        }
        
        graph.put(node, newNode);
        
        for (Node neighbor : node.neighbors) {
            if (graph.containsKey(neighbor)) {
                newNode.neighbors.add(graph.get(neighbor));    
            } else {
                newNode.neighbors.add(dfs(graph, neighbor));
            }
        }
        
        return newNode;
    }
}